File: 백준/Silver/1780. 종이의 개수/종이의 개수.cc
#include <bits/stdc++.h>
using namespace std;


int N;
int paper[2500][2500];
int one = 0, zero = 0, minusOne = 0;

bool correctNum(int x, int y, int size) {
    int first = paper[x][y];
    for (int i = x; i < x + size; i++) {
        for (int j = y; j < y + size; j++) {
            if (paper[i][j] != first) return false;
        }
    }
    return true;
}

void func(int x, int y, int size) {
    if (correctNum(x, y, size)) {
        if (paper[x][y] == 1) one++;
        else if (paper[x][y] == 0) zero++;
        else if (paper[x][y] == -1) minusOne++;
        return;
    }

    int nextSize = size / 3;
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            func(x + i * nextSize, y + j * nextSize, nextSize);
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    cin >> N;

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            cin >> paper[i][j];
        }
    }

    func(0, 0, N);

    cout << minusOne << '\n';
    cout << zero << '\n';
    cout << one << '\n';
}

File: 백준/Silver/1780. 종이의 개수/종이의 개수.py
import sys
from typing import MutableSequence
input = sys.stdin.readline
minusOne = 0
zero = 0
one = 0
def allCollect(board, x, y, size)->bool:
    first = board[x][y]
    for a in range(x, x+size):
        for b in range(y, y+size):
            if first != board[a][b]:
                return False
    return True


def func(board: MutableSequence, x : int, y : int, size: int) -> None:
    if(allCollect(board, x, y, size) or size==1):
        global minusOne, zero, one
        if(board[x][y] == -1 ):
            minusOne +=1
            return
        if(board[x][y] == 0 ):
            zero +=1
            return
        if(board[x][y] == 1 ):
            one +=1
            return
    else:
        for i in range(3):
            for j in range(3):
                func(board, x+(size//3*i), y+(size//3*j), size//3)
        

if __name__=="__main__":
    N = int(input())

    board = [[None for _ in range(N)] for _ in range(N)]
    for i in range(N):
        board[i] = list(map(int, input().split()))

    func(board, 0, 0, N)

    print(minusOne)
    print(zero)
    print(one)